### PR TITLE
Miscellaneous internal changes

### LINKED ISSUE #255: Incorporate misc changes loosely-related to #252
## 💡 Description
Preliminary work undertaken as part of #252 which does not actually resolve it (since it has been deprioritized but the work is still useful).
 - Add internal `IllegalArgumentException` to invalid `start` and `limit` setter calls
 - Remove summaryOnly-related internal code and rework Singular/Plural API route behaviour to be more explicit

### PR DESCRIPTION
**Pull request by alexdunnjpl**
## 🗒️ Summary
Preliminary work undertaken as part of #252 which does not actually resolve it (since it has been deprioritized but the work is still useful).
 - Add internal `IllegalArgumentException` to invalid `start` and `limit` setter calls
 - Remove summaryOnly-related internal code and rework Singular/Plural API route behaviour to be more explicit

## ⚙️ Test Data and/or Report
One of the following should be included here:
See Github Actions tests

## ♻️ Related Issues
Fixes #255 
Related to #252

### CODE REVIEW

File: model/swagger.yml
Comment (tloubrieu-jpl) [inline model/swagger.yml:1153]: I'll keep that note, although data is still returned with length 0, I would like that we made obvious to the users that limit=0 is the way to get the summary only.

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (tloubrieu-jpl) [inline service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: Somehow the property name is not meaningful to me. A bit too abstract maybe.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java
Comment (tloubrieu-jpl) [inline service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java:94]: why is the identation so far on the right ?

File: service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java
Comment (tloubrieu-jpl) [inline service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java]: If I am not mistaken, I mean if I understand what you do with returnSingularDatum: we don't want to return a specific format when limit=0 or 1. The format singular or plural is defined by the end-point.

It is very likely that I don't understand what you do here, so we can discuss that later.

**Review by tloubrieu-jpl — COMMENTED** [review]
Overall I like the simplification of the code, but I would like to make sure the behavior is still what we expect/want.

Thanks Alex

File: model/swagger.yml
Comment (alexdunnjpl) [reply model/swagger.yml:1153]: Is that useful information to provide to the user?  "If you request zero data you'll get zero data" seems pointless, and specifying that you'll get just the summary (which is actually the search metadata) seems to indicate that something nonstandard occurs.  Your call, though.

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: This replaces `isSingular` (dragged down from `RequestAndResponseContext`).  The meaning is "this `URIParameters` is being used for an API route which returns a single element (ex. `/bundles/{identifier}`) rather than a collection of elements (ex. `/bundles/{identifier}/products`)".

Open to suggestions on this as you're right - it's not great.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java:94]: Probably something I missed, and I haven't got my autoformatter set up yet - will fix

File: service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java]: Previously, there was some slightly-murky (to someone who isn't already familiar with it) state checking of (values pulled from) the `URIParameters` going on to determine whether the API route was a return-element or return-collection type.

That implicit logic has been made explicit and encapsulated within the `URIParameters` object.

The logic itself has not changed meaningfully - if the setters for `start` or `limit` are called on `URIParameters` object, the flag changes from return-element (returnSingularDatum=True) to return-collection(returnSingularDatum=False).  I didn't want to overcomplicate the changes, but if this is sufficiently unclear, a potential improvement would be to change the relevant member from `boolean isSingle` to an enumerated type `routeReturnType` with values `SingleElement` and `ElementCollection` or something similar.

File: service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/model/Pds4ProductBusinessObject.java:94]: Actually this wasn't a new change - there's some mixing of spaces and tabs throughout this file.  I've reformatted that function, but can expand the scope if desired.

File: model/swagger.yml
Comment (jordanpadams) [reply model/swagger.yml:1153]: @tloubrieu-jpl ☝️

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (jordanpadams) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: @tloubrieu-jpl @alexdunnjpl ☝️

**Review by nutjob4life — APPROVED** [review]

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (tloubrieu-jpl) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: @alexdunnjpl , What about something like `expectSingularResponse` ?

File: model/swagger.yml
Comment (tloubrieu-jpl) [reply model/swagger.yml:1153]: @alexdunnjpl we wanted a way for the user to get the metadata for a set of results, beyond the pagination limits, without requesting them all. That is useful for a user who would like to know the set of properties which can be requested in this products.

The use case is:
1. user want to explore the product's of a collection (e.g. 100000s products)
2. the user needs to know which properties are used to describe these products
3. the user can apply search criteria on some of the available properties to subset the collection of products (these properties cannot be guessed by the user)

That is for step 2 that we initially defined the `summaryOnly` keyword, and then we decided to remove this keyword to use limit=0 instead. That being said it sounds like the summary does not provide the description for the full request results beyond pagination limits but only for the current page.

Comment (alexdunnjpl) [reply model/swagger.yml:1153]: @tloubrieu-jpl if I understand you correctly, it sounds like `limit=0` _is_ supposed to behave in a different/non-standard manner, then? Specifically, to provide the union of the set of properties present in each product for the given query? (That's a non-homogeneous collection of property sets, yeah? Otherwise the user could just get it with `limit=1` and extract the property set themselves)

It currently doesn't, iirc, so sounds like there's a need to make some kind of fix if that's necessary, but that almost seems like a job for a subroute like `products/properties?{someQuery}` rather than serving it implicitly in the case of `products?{someQuery}&limit=0`?  This also seems like a potentially-expensive operation in the case of large query results, not that there's any getting around that if it's needed.

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: I think it's slightly better, but I'm starting to feel like if it's worth worrying about, it's worth doing properly ([see the last bit of the comment in the next thread](https://github.com/NASA-PDS/registry-api/pull/256#discussion_r1116109721)), so unless you have an objection I think that I should 
 - make it an enum type instead of a boolean
 - put an explanatory comment in the enum type so that anyone who wonders/cares what exactly it refers to can jump to the explanation with a minimum of effort

Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: Actually, how about `singletonResultExpected`?

Communicates that the expected singularness applies to the _result_ (of the query), and I think that "singleton" makes the connection a little clearer because it either means "a unique instance" or "a set with cardinality 1", both of which are correct in this context.

**Review by tloubrieu-jpl — APPROVED** [review]
I fixed the conflict (that I introduced through an update on main branch).
I ran the integration test and I don't see an impact, before merging that would be good to have a review from @al-niessner.

**Comment by tloubrieu-jpl** [comment]
@al-niessner could you review this PR ? Thanks

**Comment by alexdunnjpl** [comment]
@al-niessner @tloubrieu-jpl still a couple of outstanding convos (above) before finalization, FYI

File: service/src/main/resources/application.properties
Comment (alexdunnjpl) [inline service/src/main/resources/application.properties]: Need to remove changes to this file before merge (note to self)

Comment (alexdunnjpl) [reply service/src/main/resources/application.properties]: resolved in 6b37541

**Review by al-niessner — APPROVED** [review]

**Comment by alexdunnjpl** [comment]
Squash when merging, with following commit message (we've lost the commit history, but at least we can retain the granular messages):

* add IllegalArgumentException throws to URIParameters setStart() and setLimit() methods
* excise summaryOnly-related behaviour
* rework Singular/Plural status of API routes
  this is more explicit than the previous approach
* remove illegal character
  presence of emoji breaks branch integration testing

File: service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/controller/URIParameters.java]: @tloubrieu-jpl tentatively resolved by 90c4fb91f1bbade1efdfc50058ab6a2945285070 (if you agree with the new naming)

File: service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java
Comment (alexdunnjpl) [reply service/src/main/java/gov/nasa/pds/api/registry/model/RequestAndResponseContext.java]: @tloubrieu-jpl unless I'm missing something, this is resolved - feel free to re-open if you have outstanding questions

File: model/swagger.yml
Comment (tloubrieu-jpl) [reply model/swagger.yml:1153]: @alexdunnjpl I was unclear, I am not looking for a non standard management of limit=0. I believe what is rather needed to be updated is the way the summary is prepared. For the use case I described above, the summary need to be made from the products matching the query (q, keyword) but ignoring the pagination parameters (start, limit). I will make a more detailed ticket for that, because all the summary properties are not equivalent.

**Pull request merged**
